ckan/lib/search/__init__.py

```diff
@@ -168,6 +193,14 @@ def rebuild(package_id=None, only_missing=False, force=False, refresh=False,
                 {'id': package_id})
             log.info('Indexing just package %r...', pkg_dict['name'])
             package_index.update_dict(pkg_dict, True)
+            # (canada fork only): background search index rebuilding
+            #TODO: upstream contrib!!
+            if in_background and task:
+                value = json.loads(task.get('value', '{}'))
+                value['indexed'] += 1
+                task['value'] = json.dumps(value)
+                task['last_updated'] = str(datetime.datetime.now(datetime.timezone.utc))
+                logic.get_action('task_status_update')({'session': model.meta.create_local_session(), 'ignore_auth': True}, task)
```

Comment:
We should separate out the background-status-updating from the index-rebuilding code by turning `rebuild` into a generator that `yield`s status updates/error messages as it works.

This way all the task-updating code can be in the caller instead of woven together with the real reindexing work being done. As a side benefit the CLI could use the same interface to present status updates to users as the indexing is running.

i.e. this whole block could be replaced with something like
```python
yield {'indexed': pkg_dict['name']}
```

and the caller could be doing something like:
```python
for status in rebuild(...):
    # task updating code based on status['indexed'] or status['error']
```

Reply:
Because the caller would still have to be a background task, we would be making a new method for the callback? The action method would still have to do some task status work to get the initial task into the DB. But yeah, this makes sense. Where is the best script to put the callback though? in `lib/jobs.py` ??

Reply:
there was some system stdout stuff in the rebuild method which has now been moved to the nicer places like the background job and the cli command with log.info and click.echo respectfully